crates/spirv-std/src/bindless.rs

```diff
@@ -0,0 +1,220 @@
+        f.debug_struct("RenderResourceHandle")
+            .field("version", &self.version())
+            .field("tag", &self.tag())
+            .field("index", unsafe { &self.index() })
+            .finish()
+    }
+}
+
+impl RenderResourceHandle {
+    pub fn new(version: u8, tag: RenderResourceTag, index: u32) -> Self {
+        let version = version as u32;
+        let tag = tag as u32;
+        let index = index as u32;
+
+        assert!(version < 64); // version wraps around, it's just to make sure invalid resources don't get another version
+        assert!(tag < 8);
+        assert!(index < (1 << 23));
+
+        Self(version << 26 | tag << 23 | index)
+    }
+
+    pub fn invalid() -> Self {
+        Self(!0)
+    }
+
+    pub fn is_valid(self) -> bool {
+        self.0 != !0
+    }
+
+    pub fn version(self) -> u32 {
+        self.0 >> 26
+    }
+
+    pub fn tag(self) -> RenderResourceTag {
+        match (self.0 >> 23) & 7 {
+            0 => RenderResourceTag::Sampler,
+            1 => RenderResourceTag::Tlas,
+            2 => RenderResourceTag::Buffer,
+            3 => RenderResourceTag::Texture,
+            invalid_tag => panic!(
+                "RenderResourceHandle corrupt: invalid tag ({})",
+                invalid_tag
+            ),
+        }
+    }
+
+    /// Safety:
+    /// This method can only safely refer to a resource if that resource
+    /// is guaranteed to exist by the caller. `RenderResourceHandle` can't
+    /// track lifetimes or keep ref-counts between GPU and CPU and thus
+    /// requires extra caution from the user.
+    pub unsafe fn index(self) -> u32 {
+        self.0 & ((1 << 23) - 1)
+    }
+
+    /// This function is primarily intended for use in a slot allocator, where the slot
+    /// needs to get re-used and it's data updated. This bumps the `version` of the
+    /// `RenderResourceHandle` and updates the `tag`.
+    pub fn bump_version_and_update_tag(self, tag: RenderResourceTag) -> Self {
+        let mut version = self.0 >> 26;
+        version = ((version + 1) % 64) << 26;
+        let tag = (tag as u32) << 23;
+        Self(version | tag | (self.0 & ((1 << 23) - 1)))
+    }
+}
+
+#[derive(Copy, Clone)]
+#[repr(transparent)]
+pub struct Buffer(RenderResourceHandle);
+
+#[allow(unused_attributes)]
+#[spirv(internal_buffer_load)]
+#[spirv_std_macros::gpu_only]
+fn internal_buffer_load<T>(buffer: u32, offset: u32) -> T {
+    unimplemented!()
+} // actually implemented in the compiler
+
+impl Buffer {
+    #[spirv_std_macros::gpu_only]
+    pub fn load<T>(self, dword_aligned_byte_offset: u32) -> T {
+        // jb-todo: figure out why this assert breaks with complaints about pointers
+        // assert!(self.0.tag() == RenderResourceTag::Buffer);
+
+        unsafe { internal_buffer_load(self.0.index(), dword_aligned_byte_offset) }
+    }
+}
+
+#[derive(Copy, Clone)]
+#[repr(transparent)]
+pub struct Texture2d(RenderResourceHandle);
```

Comment:
Texture vs. Image ... Meh, I've always felt we should've called our types Textures since that's what everybody (except Vulkan / SPIR-V) calls them, so it's what I've called them here.